File: mobile/plugins/k2-plugin/android/src/main/java/io/kaitu/k2plugin/NativeLogger.kt
package io.kaitu.k2plugin

import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Thread-safe file logger for native-layer events in the VPN service process.
 * Writes to `native.log` in the app's logs directory.
 */
object NativeLogger {
    private var logFile: File? = null
    private val lock = Any()
    private const val MAX_FILE_SIZE = 20L * 1024 * 1024 // 20 MB (was 50 MB)

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    /** Initialize logger with the logs directory. Call from K2VpnService.startVpn(). */
    fun setup(logsDir: File) {
        synchronized(lock) {
            logsDir.mkdirs()
            logFile = File(logsDir, "native.log")
        }
    }

    /** Append a log entry. Thread-safe. */
    fun log(level: String, message: String) {
        synchronized(lock) {
            val file = logFile ?: return

            // Truncate if over size limit
            if (file.exists() && file.length() > MAX_FILE_SIZE) {
                file.writeText("")
            }

            val timestamp = dateFormat.format(Date())
            val line = "[$timestamp] [${level.uppercase()}] $message\n"
            file.appendText(line)
        }
    }
}

File: mobile/plugins/k2-plugin/ios/Plugin/NativeLogger.swift
import Foundation
import os.log

/// Thread-safe file logger for native-layer events in the NE process.
/// Writes to `native.log` in the shared App Group logs directory.
/// Only initialized in PacketTunnelProvider (NE process), NOT in K2Plugin (App process).
final class NativeLogger {
    static let shared = NativeLogger()

    private let queue = DispatchQueue(label: "com.allnationconnect.anc.wgios.native-logger")
    private var fileHandle: FileHandle?
    private var logFileURL: URL?
    private let maxFileSize: UInt64 = 20 * 1024 * 1024 // 20 MB (was 50 MB)
    private let dateFormatter = ISO8601DateFormatter()

    private init() {}

    /// Initialize logger with the shared logs directory.
    /// Call from PacketTunnelProvider.startTunnel() with the App Group logs path.
    func setup(logsDir: URL) {
        queue.sync {
            let fileURL = logsDir.appendingPathComponent("native.log")
            self.logFileURL = fileURL

            if !FileManager.default.fileExists(atPath: fileURL.path) {
                FileManager.default.createFile(atPath: fileURL.path, contents: nil)
            }

            self.fileHandle = try? FileHandle(forWritingTo: fileURL)
            self.fileHandle?.seekToEndOfFile()
        }
    }

    /// Append a log entry. Thread-safe.
    func log(_ level: String, _ message: String) {
        queue.async { [weak self] in
            guard let self = self, let handle = self.fileHandle, let url = self.logFileURL else { return }

            // Check file size, truncate if over limit
            if let attrs = try? FileManager.default.attributesOfItem(atPath: url.path),
               let size = attrs[.size] as? UInt64, size > self.maxFileSize {
                handle.truncateFile(atOffset: 0)
                handle.seekToEndOfFile()
            }

            let timestamp = self.dateFormatter.string(from: Date())
            let line = "[\(timestamp)] [\(level.uppercased())] \(message)\n"
            if let data = line.data(using: .utf8) {
                handle.write(data)
            }
        }
    }

    /// Close the file handle.
    func close() {
        queue.sync {
            fileHandle?.closeFile()
            fileHandle = nil
        }
    }
}
